src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                    method.IsAccessor() ||
+                    method.IsAnonymousFunction())
+                {
+                    return false;
+                }
+
+                // Ignore event handler methods "Handler(object, MyEventArgs)"
+                if (_eventArgsType != null &&
+                    method.Parameters.Length == 2 &&
+                    method.Parameters[0].Type.SpecialType == SpecialType.System_Object &&
+                    method.Parameters[1].Type.InheritsFromOrEquals(_eventArgsType))
+                {
+                    return false;
+                }
+
+                // Ignore methods with any attributes in 'attributeSetForMethodsToIgnore'.
+                if (method.GetAttributes().Any(a => a.AttributeClass != null && _attributeSetForMethodsToIgnore.Contains(a.AttributeClass)))
+                {
+                    return false;
+                }
+
+                // Ignore methods that were used as delegates
+                if (_methodsUsedAsDelegates.ContainsKey(method))
+                {
+                    return false;
+                }
+
+                return true;
+            }
+
+            private sealed class BlockAnalyzer
+            {
+                private readonly SymbolStartAnalyzer _symbolStartAnalyzer;
+                private readonly Options _options;
+                private bool _hasDelegateCreationOrAnonymousFunction;
+                private bool _escapedDelegate;
+                private bool _hasConversionFromDelegateTypeToNonDelegateType;
+
+                private BlockAnalyzer(SymbolStartAnalyzer symbolStartAnalyzer, Options options)
+                {
+                    _symbolStartAnalyzer = symbolStartAnalyzer;
+                    _options = options;
+                }
+
+                public static void Analyze(OperationBlockStartAnalysisContext context, SymbolStartAnalyzer symbolStartAnalyzer)
+                {
+                    if (HasSyntaxErrors())
+                    {
+                        return;
+                    }
+
+                    // All operation blocks for a symbol belong to the same tree.
+                    var firstBlock = context.OperationBlocks[0];
+                    var supportsDiscard = symbolStartAnalyzer._supportsDiscard(firstBlock.Syntax.SyntaxTree);
+                    var unusedValueExpressionStatementOption = symbolStartAnalyzer._unusedValueExpressionStatementOption;
+                    var unusedValueAssignmentOption = symbolStartAnalyzer._unusedValueAssignmentOption;
+                    if (!TryGetOptions(firstBlock.Syntax.SyntaxTree, firstBlock.Language,
+                                       context.Options, supportsDiscard,
+                                       unusedValueExpressionStatementOption, unusedValueAssignmentOption,
+                                       context.CancellationToken, out var options))
+                    {
```

Comment:
this is in a static method, and seems to use nothing aobut this type, while passing along a lot of darta from a containing type.  this strongly makes it feel as if this method is in the wrong place.  for example, if it were an instance methods of symbolStartAnalyzer, it seems like it would be cleaner.

Reply:
Thanks, yep that indeed seems cleaner.

---
In reply to: [230605108](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605108) [](ancestors = 230605108)